src/OpenLoco/Input/MouseInput.cpp

```diff
@@ -590,99 +591,166 @@ namespace OpenLoco::Input
+        {
+            Ui::Windows::Construction::openWithFlags(1ULL << 31);
+            return;
+        }
+        GameCommands::setErrorTitle(StringIds::cant_remove_airport);
+        GameCommands::AirportRemovalArgs args;
+        args.pos = Pos3(pos.x, pos.y, station->baseZ() * 4);
+        if (GameCommands::do_57(GameCommands::Flags::apply, args))
+        {
+            Audio::playSound(Audio::SoundId::demolish, GameCommands::getPosition());
+        }
     }
 
     // 0x004A5CC5 TODO: Move to a better file
-    static void dockInteract(Window* main, Map::StationElement* station, const Map::Pos2 pos)
+    static void dockInteract(Map::StationElement* station, const Map::Pos2 pos)
     {
-        registers regs{};
-        regs.esi = reinterpret_cast<uint32_t>(main);
-        regs.edx = reinterpret_cast<uint32_t>(station);
-        regs.ax = pos.x;
-        regs.cx = pos.y;
-        call(0x004A5CC5, regs);
+        if (!Ui::Windows::Construction::isStationTabOpen())
+        {
+            Ui::Windows::Construction::openWithFlags(1ULL << 30);
+            return;
+        }
+        GameCommands::setErrorTitle(StringIds::cant_remove_ship_port);
+        GameCommands::PortRemovalArgs args;
+        Pos2 firstTile = pos - Map::offsets[station->multiTileIndex()];
+        args.pos = Pos3(firstTile.x, firstTile.y, station->baseZ() * 4);
+        if (GameCommands::do_61(GameCommands::Flags::apply, args))
+        {
+            Audio::playSound(Audio::SoundId::demolish, GameCommands::getPosition());
+        }
     }
 
     // 0x004BB116 TODO: Move to a better file
     static void treeInteract(Map::TreeElement* tree, const Map::Pos2 pos)
     {
-        registers regs{};
-        regs.edx = reinterpret_cast<uint32_t>(tree);
-        regs.ax = pos.x;
-        regs.cx = pos.y;
-        call(0x004BB116, regs);
+        GameCommands::setErrorTitle(StringIds::error_cant_remove_this);
+        GameCommands::TreeRemovalArgs args;
+        args.pos = Pos3(pos.x, pos.y, tree->baseZ() * 4);
+        args.elementType = tree->rawData()[0];
+        args.type = tree->treeObjectId();
+        GameCommands::do_22(GameCommands::Flags::apply, args);
     }
 
     // 0x0042D9BF TODO: Move to a better file
     static void buildingInteract(Map::BuildingElement* building, const Map::Pos2 pos)
     {
-        registers regs{};
-        regs.edx = reinterpret_cast<uint32_t>(building);
-        regs.ax = pos.x;
-        regs.cx = pos.y;
-        call(0x0042D9BF, regs);
+        GameCommands::setErrorTitle(StringIds::error_cant_remove_this);
+        GameCommands::BuildingRemovalArgs args;
+        Pos2 firstTile = pos - Map::offsets[building->multiTileIndex()];
+        args.pos = Pos3(firstTile.x, firstTile.y, building->baseZ() * 4);
+        GameCommands::do_45(GameCommands::Flags::apply, args);
     }
 
     // 0x004C4809 TODO: Move to a better file
     static void wallInteract(Map::WallElement* wall, const Map::Pos2 pos)
```

Comment:
How is this triggered? I find myself unable to right-click walls/fences to delete them.

Reply:
I have not been able to trigger this. The wall is excluded from the filter so I don't think this is reachable atm.